[llvm-cov] Coverage report HTML UI to jump between uncovered parts of code

hanickadot opened the pull request:
I replaced "jump to first uncovered line" with UI buttons and keyboard shortcut to jump between uncovered parts of code: lines (key L), branchs (key B), regions (key R).

User can also jump in reverse direction with shift+key.

hanickadot commented:
Motivation: I recently was writing tests for uncovered code in a long file and I spend a lot of time scrolling and looking for missing uncovered lines inside instantiation (to which the link "jump to first uncovered line" didn't even work.

hanickadot commented:
<details>
  <summary>previous</summary>
https://youtu.be/B48yDtQq2FE

</details>
<details>
  <summary>after</summary>
  https://www.youtube.com/watch?v=LEgrRTa5B4E
</video>

</details>

hanickadot commented:
Screenshot before:
<img width="950" alt="Screenshot 2024-06-15 at 17 35 43" src="https://github.com/llvm/llvm-project/assets/6557263/6fbd7fd8-6722-4ae9-8483-690ab08b4eb1">
[video link](https://youtu.be/B48yDtQq2FE)

Screenshot after:
<img width="956" alt="Screenshot 2024-06-15 at 17 35 06" src="https://github.com/llvm/llvm-project/assets/6557263/a5434d32-926c-4311-a6f8-c18b71e5acd0">
[video link](https://youtu.be/LEgrRTa5B4E)

hanickadot commented:
Demo here: https://tmp.cpp.fail/report/coverage/include/cthash/sha3/common.hpp.html

ornata commented:
Cool! I tested in Firefox, Safari, and Chrome and the feature works in all 3.

chapuni commented on llvm/tools/llvm-cov/SourceCoverageViewHTML.cpp:
```diff
@@ -88,13 +88,123 @@ const char *BeginHeader =
+const char *JSForCoverage =
+    R"javascript(
+
+function next_uncovered(selector, reverse, scroll_selector) {
```
Could we unify the cursor? I guess users want to jump to the prev/next *visible* issue.
For example, "Jump to the prev/next another uncovered line in sight when the cursor is navigated by branch."

hanickadot replied:
I don't know how, at least not without really big refactor in how the HTML is generated.

chapuni commented:
Could we split out `.css` and `.js` to individual include files in the future? (I don't think we should do right now)

hanickadot commented:
> Could we split out `.css` and `.js` to individual include files in the future? (I don't think we should do right now)

It is, at least for me I have split .css and .js included by filepath and not as content. Or do you mean something else?

hanickadot commented:
> Could we split out `.css` and `.js` to individual include files in the future? (I don't think we should do right now)

Oh, I wish we could just #embed it :)

hanickadot commented on llvm/tools/llvm-cov/SourceCoverageViewHTML.cpp:
```diff
@@ -487,6 +651,9 @@ Error CoveragePrinterHTML::createIndexFile(
   if (Error E = emitStyleSheet())
     return E;
 
+  if (Error E = emitJavaScript())
+    return E;
+
```
thanks, fixed!

hanickadot committed:
[llvm-cov] HTML generated code now contains JavaScript code to quickly navigate between uncovered parts of code

chapuni reviewed (approved):
Ok-maru-suisan, thanks! @hanickadot
